plugins/single_plugins/window-rules.cpp

```diff
@@ -1,307 +1,161 @@
-                exec.action(view);
-        };
+void wayfire_window_rules_t::fini()
+{
+    output->disconnect_signal("map-view", &_created);
+    output->disconnect_signal("view-maximized", &_maximized);
+    output->disconnect_signal("view-minimized", &_minimized);
+    output->disconnect_signal("view-fullscreen", &_fullscreened);
+}
 
-        return result;
+void wayfire_window_rules_t::apply(const std::string &signal, wf::signal_data_t *data)
+{
+    if (data == nullptr) {
+        return;
     }
 
-    wf::signal_callback_t created, maximized, fullscreened;
-
-    std::map<std::string, std::vector<rule_func>> rules_list;
-
-    public:
-    void init()
+    for (const auto &rule : _rules)
     {
-        auto section = wf::get_core().config.get_section("window-rules");
-        for (auto opt : section->get_registered_options())
-        {
-            auto rule = parse_add_rule(opt->get_value_str());
-            rules_list[rule.signal].push_back(rule.func);
+        auto view = get_signaled_view(data);
+        _access_interface.set_view(view);
+        _action_interface.set_view(view);
+        auto error = rule->apply(signal, _access_interface, _action_interface);
+        if (error) {
+            std::cerr << "Window-rules: Error while executing rule on " << signal << " signal." << std::endl;
         }
+    }
 
-        created = [=] (wf::signal_data_t *data)
-        {
-            for (const auto& rule : rules_list["created"])
-                rule(get_signaled_view(data));
-        };
-        output->connect_signal("map-view", &created);
+    auto bounds = _lambda_registrations->rules();
+    auto begin = std::get<0>(bounds);
+    auto end = std::get<1>(bounds);
+    while (begin != end)
+    {
+        auto registration = std::get<1>(*begin);
+        bool error = false;
 
-        maximized = [=] (wf::signal_data_t *data)
+        if (registration->access_interface == nullptr)
         {
-            auto conv = static_cast<view_tiled_signal*> (data);
-            assert(conv);
-
-            if (conv->edges != wf::TILED_EDGES_ALL)
-                return;
+            auto view = get_signaled_view(data);
+            _access_interface.set_view(view);
 
-            for (const auto& rule : rules_list["maximized"])
-                rule(conv->view);
-        };
-        output->connect_signal("view-maximized", &maximized);
+            // TODO: Remove debug/test code.
+            std::cout << "Apply rule to view: " << view->get_app_id() << ": ";
 
-        fullscreened = [=] (wf::signal_data_t *data)
-        {
-            auto conv = static_cast<view_fullscreen_signal*> (data);
-            assert(conv);
-
-            if (!conv->state || conv->carried_out)
-                return;
+            error = registration->rule_instance->apply(signal, _access_interface);
+        }
+        else {
+            error = registration->rule_instance->apply(signal, *registration->access_interface);
```

Comment:
Shouldn't you set the access interface view here too ?

Reply:
The access interface type referred to here may use another source of information then the view. If operation on the view data is desired, it is better that the plugin making the registration leaves this at nullptr. I only added the option of setting a non-standard access interface to allow plugins to use different data source type. If a custom access interface is used, Window-rules does not know anything about it. We can discuss changing this feature because we never really set hard requirements for it before.

Reply:
Hmm, I see. However, currently all signals window-rules connect to are about views, so the event is only meaningful if there is a view bound to it? Or maybe I am misunderstanding.

Reply:
On the other hand, you are right, we might want to add events like output-added in the future ...